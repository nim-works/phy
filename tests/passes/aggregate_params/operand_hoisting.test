discard """
  description: "
    Ensure that the operands of all non-call, non-path operations are
    materialized to a temporary when a succeeding expression within the
    same statement is hoisted
  "
  output: "(Error ecIllegalAccess)"
"""
(TypeDefs
  (Record 4 4
    (Field 0 (Int 4)))
  (ProcTy (Int 4) (Type 0))
  (ProcTy (Void)))
(GlobalDefs)
(ProcDefs
  (ProcDef (Type 1)
    (Locals (Type 0))
    (List
      (Block
        (Params (Local 0))
        (Return (IntVal 4)))))

  (ProcDef (Type 2)
    (Locals (Type 0) (Int 4))
    (List
      (Block (Params)
        (Drop
          (Add (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (Sub (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (Mul (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (Div (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (Mod (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (AddChck (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))
            (Local 1)))
        (Drop
          (SubChck (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))
            (Local 1)))
        (Drop
          (BitAnd (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (BitOr (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (BitXor (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (Eq (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (Lt (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (Le (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (Shl (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))
        (Drop
          (Shr (Int 4)
            (Copy (Local 1))
            (Call (Type 1) (Proc 0) (Copy (Local 0)))))

        (Blit
          (Copy (Local 1))
          (Copy (Local 1))
          (Call (Type 1) (Proc 0) (Copy (Local 0))))
        (Clear
          (Copy (Local 1))
          (Call (Type 1) (Proc 0) (Copy (Local 0))))
        (Return)))))
