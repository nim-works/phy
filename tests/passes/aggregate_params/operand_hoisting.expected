;$sexp
(TypeDefs
  (Record 4 4
    (Field 0 (Int 4)))
  (ProcTy (Int 4) (UInt 8))
  (ProcTy (Void)))
(GlobalDefs)
(ProcDefs
  (ProcDef (Type 1)
    (Locals (UInt 8))
    (List
      (Block
        (Params (Local 0))
        (Return (IntVal 4)))))
  (ProcDef (Type 2)
    (Locals (Type 0) (Int 4) (Type 0) (Int 4) (Type 0) (Int 4) (Type 0)
      (Int 4) (Type 0) (Int 4) (Type 0) (Int 4) (Type 0) (Int 4) (Type 0)
      (Int 4) (Type 0) (Int 4) (Type 0) (Int 4) (Type 0) (Int 4) (Type 0)
      (Int 4) (Type 0) (Int 4) (Type 0) (Int 4) (Type 0) (Int 4) (Type 0)
      (Int 4) (Type 0) (Int 4) (Int 4) (Type 0) (Int 4))
    (List
      (Block (Params)
        (Asgn (Local 3)
          (Copy (Local 1)))
        (Asgn (Local 2)
          (Copy (Local 0)))
        (Drop
          (Add (Int 4)
            (Copy (Local 3))
            (Call (Type 1) (Proc 0)
              (Addr (Local 2)))))
        (Asgn (Local 5)
          (Copy (Local 1)))
        (Asgn (Local 4)
          (Copy (Local 0)))
        (Drop
          (Sub (Int 4)
            (Copy (Local 5))
            (Call (Type 1) (Proc 0)
              (Addr (Local 4)))))
        (Asgn (Local 7)
          (Copy (Local 1)))
        (Asgn (Local 6)
          (Copy (Local 0)))
        (Drop
          (Mul (Int 4)
            (Copy (Local 7))
            (Call (Type 1) (Proc 0)
              (Addr (Local 6)))))
        (Asgn (Local 9)
          (Copy (Local 1)))
        (Asgn (Local 8)
          (Copy (Local 0)))
        (Drop
          (Div (Int 4)
            (Copy (Local 9))
            (Call (Type 1) (Proc 0)
              (Addr (Local 8)))))
        (Asgn (Local 11)
          (Copy (Local 1)))
        (Asgn (Local 10)
          (Copy (Local 0)))
        (Drop
          (Mod (Int 4)
            (Copy (Local 11))
            (Call (Type 1) (Proc 0)
              (Addr (Local 10)))))
        (Asgn (Local 13)
          (Copy (Local 1)))
        (Asgn (Local 12)
          (Copy (Local 0)))
        (Drop
          (AddChck (Int 4)
            (Copy (Local 13))
            (Call (Type 1) (Proc 0)
              (Addr (Local 12)))
            (Local 1)))
        (Asgn (Local 15)
          (Copy (Local 1)))
        (Asgn (Local 14)
          (Copy (Local 0)))
        (Drop
          (SubChck (Int 4)
            (Copy (Local 15))
            (Call (Type 1) (Proc 0)
              (Addr (Local 14)))
            (Local 1)))
        (Asgn (Local 17)
          (Copy (Local 1)))
        (Asgn (Local 16)
          (Copy (Local 0)))
        (Drop
          (BitAnd (Int 4)
            (Copy (Local 17))
            (Call (Type 1) (Proc 0)
              (Addr (Local 16)))))
        (Asgn (Local 19)
          (Copy (Local 1)))
        (Asgn (Local 18)
          (Copy (Local 0)))
        (Drop
          (BitOr (Int 4)
            (Copy (Local 19))
            (Call (Type 1) (Proc 0)
              (Addr (Local 18)))))
        (Asgn (Local 21)
          (Copy (Local 1)))
        (Asgn (Local 20)
          (Copy (Local 0)))
        (Drop
          (BitXor (Int 4)
            (Copy (Local 21))
            (Call (Type 1) (Proc 0)
              (Addr (Local 20)))))
        (Asgn (Local 23)
          (Copy (Local 1)))
        (Asgn (Local 22)
          (Copy (Local 0)))
        (Drop
          (Eq (Int 4)
            (Copy (Local 23))
            (Call (Type 1) (Proc 0)
              (Addr (Local 22)))))
        (Asgn (Local 25)
          (Copy (Local 1)))
        (Asgn (Local 24)
          (Copy (Local 0)))
        (Drop
          (Lt (Int 4)
            (Copy (Local 25))
            (Call (Type 1) (Proc 0)
              (Addr (Local 24)))))
        (Asgn (Local 27)
          (Copy (Local 1)))
        (Asgn (Local 26)
          (Copy (Local 0)))
        (Drop
          (Le (Int 4)
            (Copy (Local 27))
            (Call (Type 1) (Proc 0)
              (Addr (Local 26)))))
        (Asgn (Local 29)
          (Copy (Local 1)))
        (Asgn (Local 28)
          (Copy (Local 0)))
        (Drop
          (Shl (Int 4)
            (Copy (Local 29))
            (Call (Type 1) (Proc 0)
              (Addr (Local 28)))))
        (Asgn (Local 31)
          (Copy (Local 1)))
        (Asgn (Local 30)
          (Copy (Local 0)))
        (Drop
          (Shr (Int 4)
            (Copy (Local 31))
            (Call (Type 1) (Proc 0)
              (Addr (Local 30)))))
        (Asgn (Local 34)
          (Copy (Local 1)))
        (Asgn (Local 33)
          (Copy (Local 1)))
        (Asgn (Local 32)
          (Copy (Local 0)))
        (Blit
          (Copy (Local 34))
          (Copy (Local 33))
          (Call (Type 1) (Proc 0)
            (Addr (Local 32))))
        (Asgn (Local 36)
          (Copy (Local 1)))
        (Asgn (Local 35)
          (Copy (Local 0)))
        (Clear
          (Copy (Local 36))
          (Call (Type 1) (Proc 0)
            (Addr (Local 35))))
        (Return)))))
